crates/ra_hir_def/src/expr.rs

```diff
@@ -381,6 +381,7 @@ pub struct RecordFieldPat {
 pub enum Pat {
     Missing,
     Wild,
+    DotDot,
```

Comment:
Do we actually need a `DotDot` pattern, or is it just the same as `Wild`? :thinking:

Reply:
Wild is `_` and matches a single item, `..` is `DotDot` and matches 0, 1, or many items. 

We probably could use a better name though...

Reply:
They need to be distinguishable in both tuple (struct) patterns and slice patterns from my understanding, since e.g. `&[_]` and `&[..]` have entirely different meaning.

Reply:
I could rename it to `Rest`, does that sound good? I think for the `..` pattern in tuple (structs), this also makes sense.

Reply:
> `&[_]` and `&[..]` have entirely different meaning.

But isn't that difference in meaning expressed in the separation between prefix, rest and suffix pattern? I.e. in my understanding, `[_]` is "prefix [_], rest none, suffix []`, whereas `[..]` is "prefix [], rest _, suffix []".

Reply:
That is to say, the `..` pattern obviously exists in the AST, but I'm not convinced it needs to exist in the desugared HIR.

Reply:
It does for tuple patterns, no? For the type `(T, T, T)` the pattern `(x, ..)` is valid but `(x, _)` isn't.

Reply:
`..` needs to be handled specially for tuple patterns. Just adding this as a pattern wouldn't help; tuple pattern inference needs to check whether there's a `..` and act accordingly. So it makes more sense to restructure tuple patterns similarly to `Pat::Slice`. The same goes for `..` in record patterns; there it wouldn't even fit into the existing structure because there's no name. In other words, `..` isn't a pattern, it's a marker with special meaning in certain kinds of patterns. It can't stand alone.

Reply:
Note that there's no such variant in rustc either:
https://github.com/rust-lang/rust/blob/7494250106003d698579edef215d0c01810b5156/src/librustc_hir/hir.rs#L867-L920

Reply:
See https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast_lowering/struct.LoweringContext.html#method.lower_pat_slice for lowering of slice patterns and https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast_lowering/struct.LoweringContext.html#method.lower_pat_tuple for tuples.